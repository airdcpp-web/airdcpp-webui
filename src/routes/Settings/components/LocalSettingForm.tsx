//import PropTypes from 'prop-types';
import { Component } from 'react';

import LocalSettingStore from 'stores/LocalSettingStore';

import Form, { FormProps, FormSaveHandler, FormFieldChangeHandler } from 'components/form/Form';

import * as API from 'types/api';
import * as UI from 'types/ui';

import { withSaveContext, SaveContextProps } from '../decorators/SaveDecorator';
import { RouteComponentProps } from 'react-router';
import { translateForm } from 'utils/FormUtils';


export interface LocalSettingFormProps extends Omit<FormProps, 'onSave' | 'fieldDefinitions' | 'value'> {
  keys: string[];
  moduleT: UI.ModuleTranslator;
}


type Props = LocalSettingFormProps & SaveContextProps & RouteComponentProps<UI.EmptyObject>;

class LocalSettingForm extends Component<Props> {
  /*static propTypes = {
    // Form items to list
    keys: PropTypes.array.isRequired,
  };*/

  definitions: UI.FormFieldDefinition[];
  state: {
    settings: API.SettingValueMap,
  };

  constructor(props: Props) {
    super(props);

    this.definitions = translateForm(LocalSettingStore.getDefinitions(props.keys), props.moduleT);

    this.state = {
      settings: LocalSettingStore.getValues(),
    };
  }

  onSave: FormSaveHandler<UI.FormValueMap> = (changedSettingArray, allFields) => {
    LocalSettingStore.setValues(changedSettingArray);

    this.setState({
      settings: LocalSettingStore.getValues(),
    });

    return Promise.resolve();
  }

  onFieldChanged: FormFieldChangeHandler = (id, value, hasChanges) => {
    const { saveContext, onFieldChanged } = this.props;
    saveContext.onFieldChanged(id, value, hasChanges);
    if (onFieldChanged) {
      return onFieldChanged(id, value, hasChanges);
    }
  }

  render() {
    const { settings } = this.state;
    const { saveContext, keys, ...other } = this.props;
    return (
      <div className="local setting-form">
        <Form
          { ...other }
          ref={ f => saveContext.addFormRef(keys, f) }
          onSave={ this.onSave }
          fieldDefinitions={ this.definitions }
          sourceValue={ settings }
          onFieldChanged={ this.onFieldChanged }
        />
      </div>
    );
  }
}

export default withSaveContext(LocalSettingForm);