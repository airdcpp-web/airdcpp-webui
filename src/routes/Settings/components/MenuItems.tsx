import RouterMenuItemLink from '@/components/semantic/RouterMenuItemLink';

import * as UI from '@/types/ui';
import { ChildSectionType, RootSectionType, SectionBase } from '../types';
import { Location } from 'react-router';

export const sectionToUrl = (section: SectionBase, parent?: RootSectionType) => {
  if (typeof parent === 'object') {
    return `/settings/${parent.url}/${section.url}`;
  }

  return `/settings/${section.url}`;
};

export const translateSettingSectionTitle = (
  title: string,
  settingsT: UI.ModuleTranslator,
) => {
  return settingsT.translate(title, [UI.SubNamespaces.NAVIGATION]);
};

const menuItemToLinkComponent = (
  url: string,
  menuItemInfo: SectionBase,
  settingsT: UI.ModuleTranslator,
  { hasAccess }: UI.AuthenticatedSession,
) => {
  if (menuItemInfo.access && !hasAccess(menuItemInfo.access)) {
    return null;
  }

  return (
    <RouterMenuItemLink
      key={url}
      url={url}
      icon={'icon' in menuItemInfo ? `green ${menuItemInfo.icon}` : null}
    >
      {translateSettingSectionTitle(menuItemInfo.title, settingsT)}
    </RouterMenuItemLink>
  );
};

export const rootMenuItemToLinkComponent = (
  rootMenuItem: RootSectionType,
  settingsT: UI.ModuleTranslator,
  location: Location,
  session: UI.AuthenticatedSession,
) => {
  // Browsing is smoother when the child page is loaded directly
  // Don't use the child URL for currently active parent so that the route is detected as active correctly
  let url = sectionToUrl(rootMenuItem);
  if (rootMenuItem.menuItems && !location.pathname.startsWith(url)) {
    url = sectionToUrl(rootMenuItem.menuItems[0], rootMenuItem);
  }

  return menuItemToLinkComponent(url, rootMenuItem, settingsT, session);
};

export const childMenuItemToLinkComponent = (
  childMenuItem: ChildSectionType,
  parent: RootSectionType | undefined,
  settingsT: UI.ModuleTranslator,
  session: UI.AuthenticatedSession,
) => {
  const url = sectionToUrl(childMenuItem, parent);
  return menuItemToLinkComponent(url, childMenuItem, settingsT, session);
};

export const findMainSection = (
  mainSection: string | undefined,
  rootMenuItems: RootSectionType[],
) => {
  const item = rootMenuItems.find((rootMenuItem) => rootMenuItem.url === mainSection);
  return item || rootMenuItems[0];
};

export const findChildSection = (
  childSection: string | undefined,
  mainSection: RootSectionType,
) => {
  const item = [...mainSection.menuItems, ...(mainSection.advancedMenuItems || [])].find(
    (childMenuItem) => childMenuItem.url === childSection,
  );

  return item || mainSection.menuItems[0];
};
