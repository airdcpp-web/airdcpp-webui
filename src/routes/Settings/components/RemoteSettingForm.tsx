import { useContext } from 'react';
import SettingConstants from 'constants/SettingConstants';

import DataProviderDecorator, {
  DataProviderDecoratorChildProps,
} from 'decorators/DataProviderDecorator';
import SocketService from 'services/SocketService';

import Form, {
  FormProps,
  FormSaveHandler,
  FormFieldChangeHandler,
} from 'components/form/Form';

import * as API from 'types/api';
import * as UI from 'types/ui';
import { useLocation } from 'react-router';
import { SettingSaveContext, getSettingFormId } from '../effects/useSettingSaveContext';

export interface RemoteSettingFormProps
  extends Omit<FormProps, 'onSave' | 'value' | 'fieldDefinitions' | 'location'> {
  keys: string[];
  onSettingValuesReceived?: (settings: API.SettingValueMap) => void;
  valueMode?: API.SettingValueMode;
}

interface RemoteSettingFormDataProps extends DataProviderDecoratorChildProps {
  settings: API.SettingValueMap;
  fieldDefinitions: UI.FormFieldDefinition[];
}

type Props = RemoteSettingFormProps & RemoteSettingFormDataProps;

const RemoteSettingForm: React.FC<Props> = ({
  onFieldChanged,
  refetchData,
  keys,
  fieldDefinitions,
  settings,
  ...other
}) => {
  const saveContext = useContext(SettingSaveContext)!;
  const location = useLocation();

  const refetchValues = () => {
    refetchData(['settings']);
  };

  const onSave: FormSaveHandler<UI.FormValueMap> = (changedValues) => {
    if (Object.keys(changedValues).length === 0) {
      return Promise.resolve();
    }

    return SocketService.post(SettingConstants.ITEMS_SET_URL, changedValues).then(
      refetchValues,
    );
  };

  const handleFieldChanged: FormFieldChangeHandler = (id, value, hasChanges) => {
    saveContext.onFieldChanged(id, value, hasChanges);
    if (onFieldChanged) {
      return onFieldChanged(id, value, hasChanges);
    }
  };

  return (
    <div className="remote setting-form">
      <Form
        {...other}
        id={getSettingFormId(keys)}
        ref={(f) => saveContext.addFormRef(keys, f)}
        onSave={onSave}
        fieldDefinitions={fieldDefinitions}
        sourceValue={settings}
        onFieldChanged={handleFieldChanged}
        location={location}
      />
    </div>
  );
};

export default DataProviderDecorator<RemoteSettingFormProps, RemoteSettingFormDataProps>(
  RemoteSettingForm,
  {
    urls: {
      fieldDefinitions: ({ keys }) =>
        SocketService.post(SettingConstants.ITEMS_DEFINITIONS_URL, { keys }),
      settings: async ({
        keys,
        onSettingValuesReceived,
        valueMode = API.SettingValueMode.CURRENT,
      }) => {
        const settings: API.SettingValueMap = await SocketService.post(
          SettingConstants.ITEMS_GET_URL,
          {
            keys,
            value_mode: valueMode,
          },
        );

        if (!!onSettingValuesReceived) {
          onSettingValuesReceived(settings);
        }

        return settings;
      },
    },
  },
);
