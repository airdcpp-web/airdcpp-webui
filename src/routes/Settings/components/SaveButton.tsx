import { useState } from 'react';
import Button from '@/components/semantic/Button';

import classNames from 'classnames';

import * as API from '@/types/api';
import * as UI from '@/types/ui';
import { SettingSaveState } from '../effects/useSettingSaveContext';
import { useSession } from '@/context/SessionContext';

export interface SaveButtonProps {
  saveState: SettingSaveState;
  className?: string;
  settingsT: UI.ModuleTranslator;
}

const SaveButton: React.FC<SaveButtonProps> = ({ saveState, settingsT, className }) => {
  const { local, handleSave, hasChanges } = saveState;
  const [saving, setSaving] = useState(false);
  const { hasAccess } = useSession();

  const onClick = async () => {
    setSaving(true);

    try {
      await handleSave();
    } finally {
      setSaving(false);
    }
  };

  const hasEditAccess: boolean = local || hasAccess(API.AccessEnum.SETTINGS_EDIT);

  let title;
  if (!hasEditAccess) {
    title = settingsT.translate('No save permission');
  } else {
    title = settingsT.translate(hasChanges ? 'Save changes' : 'No unsaved changes');
  }

  return (
    <Button
      className={classNames('save', className)}
      caption={title}
      icon={hasEditAccess && hasChanges ? 'green checkmark' : null}
      loading={saving}
      disabled={!hasChanges || !hasEditAccess}
      onClick={onClick}
    />
  );
};

export default SaveButton;
