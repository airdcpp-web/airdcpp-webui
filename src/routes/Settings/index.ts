import IconConstants from 'constants/IconConstants';

import * as API from 'types/api';
import * as UI from 'types/ui';

export const SettingRoutes: UI.RouteItem = {
  title: 'Settings',
  path: '/settings',
  // matchPath: '/settings/:mainSection?/:childSection?',
  icon: IconConstants.SETTINGS,
  access: API.AccessEnum.SETTINGS_VIEW,
  lazy: async () => {
    const { default: Component } = await import(
      /* webpackChunkName: "settings" */ './components/Settings'
    );
    return {
      Component,
    };
  },
};
