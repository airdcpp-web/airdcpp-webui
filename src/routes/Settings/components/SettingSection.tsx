import * as React from 'react';
import classNames from 'classnames';

import { usingMobileLayout } from 'utils/BrowserUtils';

import SettingsSideMenuLayout from './SettingsSideMenuLayout';
import SettingsTopMenuLayout from './SettingsTopMenuLayout';

import * as UI from 'types/ui';

import { childMenuItemToLinkComponent, rootMenuItemToLinkComponent } from './MenuItems';
import { useLocation } from 'react-router';
import { ChildSectionType, RootSectionType } from '../types';
import {
  SettingSaveContext,
  useSettingSaveContext,
} from '../effects/useSettingSaveContext';
import Message from 'components/semantic/Message';
import IconConstants from 'constants/IconConstants';
import SaveButton from './SaveButton';
import { useSession } from 'context/SessionContext';

export interface SettingRootSectionProps {
  settingsT: UI.ModuleTranslator;
  rootMenuItems: RootSectionType[];
  selectedRootMenuItem: RootSectionType;
  selectedChildMenuItem: ChildSectionType;
  children: React.ReactNode;
}

export type SettingSectionProps = SettingRootSectionProps;

const getMessage = (childMenuItem: ChildSectionType, settingsT: UI.ModuleTranslator) => {
  return (
    childMenuItem.local && (
      <Message
        description={settingsT.t(
          'browserSpecificNote',
          'Settings listed on this page are browser-specific',
        )}
        icon={IconConstants.INFO}
      />
    )
  );
};

const SettingSection: React.FC<SettingSectionProps> = (props) => {
  const login = useSession();
  const {
    rootMenuItems,
    selectedRootMenuItem,
    settingsT,
    children,
    selectedChildMenuItem,
  } = props;

  const location = useLocation();
  const { saveState, saveContext } = useSettingSaveContext({
    selectedChildMenuItem,
    settingsT,
  });

  const SectionLayoutComponent =
    usingMobileLayout() || window.innerWidth < 1000
      ? SettingsTopMenuLayout
      : SettingsSideMenuLayout;

  const { menuItems, advancedMenuItems } = selectedRootMenuItem;

  const contentClassname = classNames(
    'section-content',
    selectedRootMenuItem.url,
    selectedChildMenuItem.url,
  );

  const menu = {
    rootMenuItems: rootMenuItems.map((item) =>
      rootMenuItemToLinkComponent(item, settingsT, location, login),
    ),
    childMenuItems: menuItems.map((item) =>
      childMenuItemToLinkComponent(item, selectedRootMenuItem, settingsT, login),
    ),
    childAdvancedMenuItems: advancedMenuItems?.map((item) =>
      childMenuItemToLinkComponent(item, selectedRootMenuItem, settingsT, login),
    ),
  };

  const getSaveButton = (className: string) => {
    if (selectedChildMenuItem.noSave) {
      return null;
    }

    return (
      <SaveButton settingsT={settingsT} saveState={saveState} className={className} />
    );
  };

  return (
    <SettingSaveContext.Provider value={saveContext}>
      <SectionLayoutComponent
        settingsT={settingsT}
        location={location}
        getSaveButton={getSaveButton}
        message={getMessage(selectedChildMenuItem, settingsT)}
        contentClassname={contentClassname}
        menu={menu}
        selectedChildMenuItem={selectedChildMenuItem}
        selectedRootMenuItem={selectedRootMenuItem}
      >
        {children}
      </SectionLayoutComponent>
    </SettingSaveContext.Provider>
  );
};

export default SettingSection;
