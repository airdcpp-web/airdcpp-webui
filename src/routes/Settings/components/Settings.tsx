import * as UI from 'types/ui';

import '../style.css';

import Profile from '../routes/Profile';
import Connectivity from '../routes/Connectivity';
import SpeedLimits from '../routes/SpeedLimits';
import Downloads from '../routes/Downloads';
import Sharing from '../routes/Sharing';
import View from '../routes/View';
import About from '../routes/About';
import Extensions from '../routes/Extensions';
import System from '../routes/System';

import { ChildSectionType, RootSectionType } from '../types';
import { Navigate, Route, Routes } from 'react-router-dom';
import SettingSection from './SettingSection';
import { getModuleT, getSubModuleT } from 'utils/TranslationUtils';
import { useTranslation } from 'react-i18next';
import { camelCase } from 'lodash';

const SettingsMenu = [
  Profile,
  Connectivity,
  SpeedLimits,
  Downloads,
  Sharing,
  View,
  About,
  Extensions,
  System,
];

export interface SettingsProps {}

const childToRoute = (
  section: ChildSectionType,
  parent: RootSectionType,
  settingsT: UI.ModuleTranslator
) => {
  const url = `${section.url}/*`;
  return (
    <Route
      key={url}
      path={url}
      element={
        <SettingSection
          settingsT={settingsT}
          selectedRootMenuItem={parent}
          rootMenuItems={SettingsMenu}
          selectedChildMenuItem={section}
        >
          <section.component
            settingsT={settingsT}
            moduleT={getSubModuleT(settingsT, camelCase(section.url))}
          />
        </SettingSection>
      }
    />
  );
};

const rootToRoute = (rootSection: RootSectionType, settingsT: UI.ModuleTranslator) => {
  return (
    <Route key={rootSection.url} path={rootSection.url}>
      {rootSection.menuItems.map((child) => childToRoute(child, rootSection, settingsT))}
      {rootSection.advancedMenuItems?.map((child) =>
        childToRoute(child, rootSection, settingsT)
      )}
      <Route index element={<Navigate to={rootSection.menuItems[0].url} />} />
    </Route>
  );
};

// Only to pass menu items to the decorated component
const Settings: React.FC<SettingsProps> = (props) => {
  const { t } = useTranslation();
  const settingsT = getModuleT(t, UI.Modules.SETTINGS);

  return (
    <div className="ui segment settings-layout">
      <Routes>
        <Route index element={<Navigate to={SettingsMenu[0].url} />} />
        {SettingsMenu.map((rootMenuItem) => rootToRoute(rootMenuItem, settingsT))}
      </Routes>
    </div>
  );
};

export default Settings;
