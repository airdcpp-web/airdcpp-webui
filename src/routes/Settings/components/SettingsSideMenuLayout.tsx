import * as React from 'react';
import classNames from 'classnames';

import LayoutHeader from 'components/semantic/LayoutHeader';
import { SettingSectionChildProps } from 'routes/Settings/components/SettingSection';
import { translateSettingSectionTitle } from './MenuItems';
import Icon from 'components/semantic/Icon';

type SideMenuProps = Pick<
  SettingSectionChildProps,
  'menuItems' | 'advancedMenuItems' | 'settingsT'
>;

const SideChildSectionMenu: React.FC<SideMenuProps> = ({
  menuItems,
  advancedMenuItems,
  settingsT,
}) => {
  return (
    <div className="three wide column menu-column">
      <div className="ui vertical secondary menu">
        {menuItems}
        {!!advancedMenuItems && (
          <div>
            <div className="item header">{settingsT.translate('Advanced')}</div>
            <div className="menu">{advancedMenuItems}</div>
          </div>
        )}
      </div>
    </div>
  );
};

type TopMenuProps = Pick<SettingSectionChildProps, 'parentMenuItems'>;

const TopRootSectionMenu: React.FC<TopMenuProps> = ({ parentMenuItems }) => (
  <div className="ui secondary pointing menu settings top-menu">{parentMenuItems}</div>
);

// eslint-disable-next-line max-len
type ContentProps = React.PropsWithChildren<
  Pick<
    SettingSectionChildProps,
    | 'contentClassname'
    | 'currentMenuItem'
    | 'parent'
    | 'saveButton'
    | 'message'
    | 'settingsT'
  >
>;

const Content: React.FC<ContentProps> = ({
  contentClassname,
  currentMenuItem,
  parent,
  saveButton,
  children,
  message,
  settingsT,
}) => (
  <div className={classNames('thirteen wide column', contentClassname)}>
    <div className="ui segment">
      <LayoutHeader
        title={translateSettingSectionTitle(currentMenuItem.title, settingsT)}
        icon={<Icon color="green" icon={parent!.icon} />}
        rightComponent={saveButton}
      />
      <div className="options">
        {message}
        {children}
      </div>
    </div>
  </div>
);

const SideMenuLayout: React.FC<SettingSectionChildProps> = ({ children, ...other }) => (
  <div className="full">
    <TopRootSectionMenu {...other} />
    <div id="setting-scroll-context" className="ui segment grid main">
      <SideChildSectionMenu {...other} />
      <Content {...other}>{children}</Content>
    </div>
  </div>
);

export default SideMenuLayout;
