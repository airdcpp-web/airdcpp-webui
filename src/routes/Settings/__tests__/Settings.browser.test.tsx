import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';

import { renderDataRoutes } from '@/tests/render/test-renderers';

import { initCommonDataMocks } from '@/tests/mocks/mock-data-common';
import SettingConstants from '@/constants/SettingConstants';
import {
  clickButton,
  expectRequestToMatchSnapshot,
  waitForData,
  waitForUrl,
} from '@/tests/helpers/test-helpers';

import { waitFor } from '@testing-library/dom';
import { getMockServer, MockServer } from '@/tests/mocks/mock-server';
import Settings, { RootSettingSections } from '../components/Settings';
import { sectionToUrl } from '../components/MenuItems';
import {
  SettingUserPageDefinitionsResponse,
  SettingUserPageValuesResponse,
} from '@/tests/mocks/api/settings';
import { setInputFieldValues, setupUserEvent } from '@/tests/helpers/test-form-helpers';

describe('Settings layout', () => {
  let server: MockServer;
  const getSocket = async () => {
    const commonData = await initCommonDataMocks(server);

    const onGetValues = vi.fn();
    server.addRequestHandler(
      'POST',
      SettingConstants.ITEMS_GET_URL,
      SettingUserPageValuesResponse,
      onGetValues,
    );

    server.addRequestHandler(
      'POST',
      SettingConstants.ITEMS_DEFINITIONS_URL,
      SettingUserPageDefinitionsResponse,
    );

    const onSave = vi.fn();
    server.addRequestHandler('POST', SettingConstants.ITEMS_SET_URL, undefined, onSave);

    return { commonData, server, onSave, onGetValues };
  };

  const renderLayout = async () => {
    const { commonData, server, ...other } = await getSocket();

    const SettingsLayoutTest = () => {
      return <Settings />;
    };

    const routes = [
      {
        path: '/settings/*',
        Component: SettingsLayoutTest,
      },
    ];

    const renderData = renderDataRoutes(routes, commonData, {
      routerProps: { initialEntries: ['/settings'] },
    });

    return { ...renderData, ...other, ...commonData };
  };

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
  });

  test('should save values', async () => {
    const renderData = await renderLayout();
    const { getByRole, router, queryByText, onSave, onGetValues } = renderData;

    await waitForUrl(
      sectionToUrl(RootSettingSections[0].menuItems[0], RootSettingSections[0]),
      router,
    );

    await waitForData('Loading data...', queryByText);
    expect(onGetValues).toBeCalledTimes(1);

    const userEvent = setupUserEvent();
    await setInputFieldValues(
      { ...renderData, userEvent },
      {
        Nick: 'New user name',
        'E-Mail (optional)': 'email@test.com',
      },
    );

    await clickButton('Save changes', getByRole);

    await waitFor(() => expect(onSave).toBeCalled());
    expectRequestToMatchSnapshot(onSave);

    await waitFor(() => expect(onGetValues).toBeCalledTimes(2));
  });
});
