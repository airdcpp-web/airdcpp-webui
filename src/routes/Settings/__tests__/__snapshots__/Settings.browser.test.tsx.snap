// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Settings layout > should save values 1`] = `
{
  "data": {
    "email": "email@test.com",
    "nick": "New user name",
  },
  "method": "POST",
  "path": "settings/set",
}
`;
